### ISSUE TITLE
Support options dependent: :destroy in Schema's has_many and belongs_to

### ISSUE THREAD

**Issue by foxnewsnetwork** [elixir-lang/ecto#655]
Ecto has a lot of influences from ActiveRecord, and one of the handiest things ActiveRecord had was:

```ruby
class School
  has_many :students, dependent: :destroy
end
```
which provided a good and conventional way of deleting all the `students` after the `school` is destroyed. It would be great if we could also have this feature in Ecto's `has_many` (and `belongs_to`) macros

As per @josevalim 's instructions from this thread in the google group:
https://groups.google.com/forum/#!topic/elixir-ecto/RLJ3f3yETdQ

**Comment by gjaldon** [comment]
I'm on this!

**Comment by josevalim** [comment]
Please hold on. :) The schema does not know anything about callbacks. We need to discuss if we are going to change that or if we are going to support this some other way.

**Comment by gjaldon** [comment]
I'm roughly thinking to do it this way:

1. Put the name of the association in a module attribute that represents all associations tagged with `dependent: :destroy`
2. Expose this module attribute with a function like `__schema__(:destroys)` (not sure about the name)
3. In `Ecto.Repo.Model.delete/4`, check if `model.__schema__(:destroys)` has anything
4. If it has, then iterate through it and get the corresponding assoication through `model.__schema__(:association, name)` and call `Ecto.Repo.Queryable.delete_all/4` passing the assoc module, the assoc_key and the id of the model so I'll get a query like `from(a in AssocModule, where: field(a, ^assoc_key) == ^model_id)`

What do you think?

**Comment by josevalim** [comment]
That's one way to go but callbacks would be conceptually cleaner because we wouldn't need to push more logic into the repositories.

**Comment by gjaldon** [comment]
So if we support callbacks inside schema blocks, we can just have a macro that expands to `after_delete :destroy_dependents` if `dependent: :destroy` opt is passed? That's easier to wrap one's head around.

How will callbacks defined inside schema differ from those defined outside?

**Comment by josevalim** [comment]
I talked to @ericmj, here is how we could implement this. We will have `Ecto.Model.Dependent` that will register a `@before_compile` hook and in the before compile check the defined associations and register the proper callbacks. All we need to do is to store the dependent information in the association reflection. Also, we shouldn't call it `destroy` because there is no `destroy` in Ecto. I think it should be `:delete_each` (fetches and delete) and `delete_all` (deletes without fetching and therefore without running callbacks).

**Comment by foxnewsnetwork** [comment]
@josevalim hey, that's a good and clean way of implementing it; really cool. But how about `:fetch_and_delete` and `:delete_without_fetch`? It's a little longer, yes, but that way, it's a little more clear to us end users what Ecto will be doing.

**Comment by josevalim** [comment]
Good point. We have possibly four options: `nothing`, `fetch_and_delete`, `nilify_all` (sets the foreign key to nil with an update_all query)  and `delete_all`. We don't need to support them all, just pointing the names. What do you think?

**Comment by foxnewsnetwork** [comment]
I think that's a great idea.

**Comment by gjaldon** [comment]
Ok copy that. Working on this now.

**Comment by zmoshansky** [comment]
Maybe a little late to the party, but what about `ON DELETE CASCADE` at the (SQL) DB level for the `delete_all`? Related to #697.

**Comment by josevalim** [comment]
@zmoshansky we should support both because not all DBs may have this functionality.

**Comment by gjaldon** [comment]
Btw, @josevalim all our previous examples were `after_delete` callbacks but working on it now, it seems these should all be `before_delete` callbacks so that the constraints errors won't be raised. 

As an update, I already have `:fetch_and_delete` and `:delete_all` callbacks done. Just waiting on `Repo.update_all/2` to be changed into a function for `:nilify_all` to work.

**Comment by josevalim** [comment]
Awesome!

**Comment by josevalim** [comment]
@gjaldon The new update_all is in master!

**Comment by gjaldon** [comment]
@josevalim thanks for the update! Will finish this up within the next few days :)

**Comment by josevalim** [comment]
Closing in favor of the excellent PR.

**Comment by gjaldon** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by foxnewsnetwork** [comment]
awesome, you guys rock
